src/main/java/org/spongepowered/common/service/SpongeServiceProvider.java

```diff
@@ -0,0 +1,278 @@
+            }
+        }
+
+        // Ta-da.
+        return new SpongeServiceProvider(services.build());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    private static <T> Registration<T> createRegistration(final Class<T> clazz, final Supplier<?> supplier, final PluginContainer container) {
+        final T impl = (T) supplier.get();
+        if (impl == null) {
+            return null;
+        }
+        return new Registration<>(clazz, (T) supplier.get(), container);
+    }
+
+    @Nullable
+    private static <T> Registration<T> getSpecificRegistration(final PluginContainer container, final Class<T> service) {
+        final ProvideServiceEventImpl<T> event =
+                new ProvideServiceEventImpl<>(SpongeCommon.getCauseStackManager().getCurrentCause(), TypeToken.of(service));
+
+        // This is the actual query - a generic event.
+        ((SpongeEventManager) Sponge.getEventManager()).post(event, container);
+        if (event.getSuggestion() != null) {
+            try {
+                return new Registration<>(service, event.getSuggestion().get(), container);
+            } catch (final Throwable e) { // if the service can't be created
+                SpongeCommon.getLogger().error("Could not create service {} from plugin {}.",
+                        service.getSimpleName(),
+                        container.getMetadata().getId(),
+                        e);
+                return null;
+            }
+        }
+        SpongeCommon.getLogger().error("Could not create service {} from plugin {}, no service was provided.", service.getSimpleName(),
+                container.getMetadata().getId());
+        return null;
+    }
+
+    // --
+
+    private final Map<Class<?>, Registration<?>> services;
+
+    public SpongeServiceProvider(final Map<Class<?>, Registration<?>> services) {
+        this.services = ImmutableMap.copyOf(services);
+    }
+
+    @Override
+    @NonNull
+    @SuppressWarnings("unchecked")
+    public <T> Optional<T> provide(@NonNull final Class<T> serviceClass) {
+        final Registration<T> registration = (Registration<T>) this.services.get(serviceClass);
+        if (registration != null) {
+            return Optional.of(registration.service());
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    @NonNull
+    @SuppressWarnings("unchecked")
+    public <T> Optional<ServiceRegistration<T>> getRegistration(@NonNull final Class<T> serviceClass) {
+        return Optional.ofNullable((ServiceRegistration<T>) this.services.get(serviceClass));
+    }
+
+    @NonNull
+    @SuppressWarnings("unchecked")
+    private <T> T provideUnchecked(final Class<T> serviceClass) {
+        final Registration<T> registration = (Registration<T>) this.services.get(serviceClass);
+        if (registration != null) {
+            return registration.service();
+        }
+        throw new IllegalStateException("");
```

Comment:
`"No provider is registered for the service '" + service.getName() + "'", service"` (this is the current errore message).

Shouldn't this thrown a `ProvisioningException` ?

Reply:
It's probably worth noting that this is a private method, so the rationale was simply that this never should happen. So ISE is fine - the object is in an illegal state if this is called and the service isn't there.

Given that exception does not exist, no.

Reply:
I missed the exception was remove in the API PR